src/tests/p11test/p11test_case_wrap.c

```diff
@@ -37,6 +38,103 @@ strip_pkcs7_padding(const unsigned char *message, unsigned long message_length,
 }
 
+static int
+check_encrypt_decrypt_secret(CK_BYTE *plain_key, CK_ULONG plain_key_len, test_cert_t *pkcs11_key,
+		token_info_t *info)
+{
+	CK_BYTE iv[] = {0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+			0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f};
+	EVP_CIPHER_CTX *ctx;
+	const EVP_CIPHER *cipher = NULL;
+	unsigned char plaintext[42] = {
+		0x42, 0x43, 0x44, 0x45, 0x46, 0x47, 0x48, 0x49,
+		0x4a, 0x4b, 0x4c, 0x4d, 0x4e, 0x4f, 0x50, 0x51,
+		0x52, 0x53, 0x54, 0x55, 0x56, 0x57, 0x58, 0x59,
+		0x5a, 0x5b, 0x5c, 0x5d, 0x5e, 0x5f, 0x60, 0x61,
+		0x62, 0x63, 0x64, 0x65, 0x66, 0x67, 0x68, 0x69,
+		0x6a, 0x6b,
+	};
+	int plaintext_len = sizeof(plaintext);
+	unsigned char ciphertext[100];
+	int ciphertext_len = sizeof(ciphertext);
+	test_mech_t aes_mech = {.mech = CKM_AES_CBC, .params = &iv, .params_len = sizeof(iv)};
+	unsigned char *check = NULL;
+	int check_len = 0;
+	int rv, len;
+
+	if (pkcs11_key->key_type != CKK_AES) {
+		fprintf(stderr, "  AES supported only\n");
+		return 1;
+	}
+
+	/* First, do the encryption dance with OpenSSL */
+	ctx = EVP_CIPHER_CTX_new();
+	if (ctx == NULL) {
+		EVP_CIPHER_CTX_free(ctx);
+		fprintf(stderr, "  EVP_CIPHER_CTX_new failed\n");
+		return -1;
+	}
+
+	if (plain_key_len == 32) {
+		cipher = EVP_aes_256_cbc();
+	} else if (plain_key_len == 16) {
+		cipher = EVP_aes_128_cbc();
+	} else {
+		EVP_CIPHER_CTX_free(ctx);
+		fprintf(stderr, "  Invalid key length %lu", plain_key_len);
+		return -1;
+	}
+	rv = EVP_EncryptInit_ex(ctx, cipher, NULL, plain_key, iv);
+	if (rv != 1) {
+		EVP_CIPHER_CTX_free(ctx);
+		fprintf(stderr, "  EVP_EncryptInit_ex failed\n");
+		return -1;
+	}
+	rv = EVP_EncryptUpdate(ctx, ciphertext, &ciphertext_len, plaintext, plaintext_len);
+	if (rv != 1) {
+		EVP_CIPHER_CTX_free(ctx);
+		fprintf(stderr, "  EVP_EncryptUpdate failed\n");
+		return -1;
+	}
+	rv = EVP_EncryptFinal_ex(ctx, ciphertext + ciphertext_len, &len);
+	EVP_CIPHER_CTX_free(ctx);
+	if (rv != 1) {
+		fprintf(stderr, "  EVP_EncryptFinal_ex failed\n");
+		return -1;
+	}
+	ciphertext_len += len;
+	/* Now, decrypt with the PKCS#11 */
+	check_len = decrypt_message(pkcs11_key, info, ciphertext, ciphertext_len, &aes_mech, &check);
+	if (check_len < 0) {
+		fprintf(stderr, "  Cannot decrypt message\n");
+		return -1;
+	}
+
+	check_len = strip_pkcs7_padding(check, check_len, 16);
+	if (check_len <= 0) {
+		free(check);
+		fprintf(stderr, "  Failed to strip PKCS#7 padding\n");
+		return -1;
+	}
+	if (check_len == plaintext_len && memcmp(plaintext, check, plaintext_len) == 0) {
+		free(check);
+		return 0;
+	}
+	/* else error */
+	fprintf(stderr, "  Decrypted message does not match (%d, %d)\n", check_len, plaintext_len);
+	fprintf(stderr, "\nplaintext:\n");
+	for (int i = 0; i < plaintext_len; i++) {
+		fprintf(stderr, ":%x", plaintext[i]);
+	}
+	fprintf(stderr, "\ncheck:\n");
+	for (int i = 0; i < check_len; i++) {
+		fprintf(stderr, ":%x", check[i]);
+	}
+	fprintf(stderr, "\n");
+	free(check);
+	return -1;
+}
+
```

Comment:
I would maybe add here note about the return values (as by `test_secret_sign_verify`).

Reply:
Thanks. Added a note and resolved the strict build issues reported.

Reply:
Actually I meant the comment for `test_wrap`, as there `0` stands for ok/skip and `1` for error (sorry for misleading comment).

Reply:
Done too.